corehq/apps/es/__init__.py

```diff
@@ -43,14 +46,22 @@
 user_adapter = users.user_adapter
 case_search_bha_adapter = case_search_bha.case_search_bha_adapter
 
-CANONICAL_NAME_ADAPTER_MAP = {
-    app_adapter.canonical_name: app_adapter,
-    case_adapter.canonical_name: case_adapter,
-    case_search_bha_adapter.canonical_name: case_search_bha_adapter,
-    case_search_adapter.canonical_name: case_search_adapter,
-    domain_adapter.canonical_name: domain_adapter,
-    form_adapter.canonical_name: form_adapter,
-    group_adapter.canonical_name: group_adapter,
-    sms_adapter.canonical_name: sms_adapter,
-    user_adapter.canonical_name: user_adapter,
-}
+
+@memoized
+def canonical_name_adapter_map():
+    """
+    Due to custom indices created in SaaS environments, we need this to by dynamic.
+    """
+    adapter_map = {
+        app_adapter.canonical_name: app_adapter,
+        case_adapter.canonical_name: case_adapter,
+        case_search_adapter.canonical_name: case_search_adapter,
+        domain_adapter.canonical_name: domain_adapter,
+        form_adapter.canonical_name: form_adapter,
+        group_adapter.canonical_name: group_adapter,
+        sms_adapter.canonical_name: sms_adapter,
+        user_adapter.canonical_name: user_adapter,
+    }
+    if settings.IS_SAAS_ENVIRONMENT:
```

Comment:
Unrelated to the changes here, this setting name feels wrong for gating a feature with this level of specificity. I think we should have a setting like `HAS_BHA_CASE_SEARCH_INDEX`. Any other conditional indexes would be similarly gated by a specific flag (although I'm not advocating that we have any others, much the contrary #GAorBust!).